### PR TITLE
Treecollecting

### PR DESCRIPTION
**Pull request by dawedawe**
This adds infrastructure to process the AST and TAST in a hopefully easier way for analyzer authors.

### CODE REVIEW

File: src/FSharp.Analyzers.SDK/ASTCollecting.fsi [hunk 1]
```diff
@@ -0,0 +1,74 @@
+namespace FSharp.Analyzers.SDK
+
+#nowarn "1182"
```
Comment (nojaf) [inline src/FSharp.Analyzers.SDK/ASTCollecting.fsi:3]: Can probably be dropped in the signature file.

File: src/FSharp.Analyzers.SDK/ASTCollecting.fsi [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+module ASTCollecting =
+    open FSharp.Compiler.Syntax
+    /// A pattern that collects all attributes from a `SynAttributes` into a single flat list
+    val (|AllAttrs|): attrs: SynAttributes -> SynAttribute list
```
Comment (nojaf) [inline src/FSharp.Analyzers.SDK/ASTCollecting.fsi:8]: Should these active patterns be exposed?

File: src/FSharp.Analyzers.SDK/ASTCollecting.fsi [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    /// An recursive pattern that collect all sequential expressions to avoid StackOverflowException
+    val (|Sequentials|_|): e: SynExpr -> SynExpr list option
+    val (|ConstructorPats|): SynArgPats -> SynPat list
+    /// A pattern that collects all patterns from a `SynSimplePats` into a single flat list
+    val (|AllSimplePats|): pats: SynSimplePats -> SynSimplePat list
+
+    type SyntaxCollectorBase =
```
Comment (nojaf) [inline src/FSharp.Analyzers.SDK/ASTCollecting.fsi:15]: This missing some good API documentation.
Perhaps even a `*.fsx` page for the docs to showcase the usage.

**Review by nojaf — COMMENTED** [review]
Good stuff! I did some nit-picking from afar. Will review properly later this week!

File: src/FSharp.Analyzers.SDK/TASTCollecting.fs [hunk 4]
```diff
@@ -0,0 +1,137 @@
+namespace FSharp.Analyzers.SDK
+
+open FSharp.Compiler.Symbols
+open FSharp.Compiler.Text
+open FSharp.Compiler.Symbols.FSharpExprPatterns
+
+module TASTCollecting =
+
+    type TypedTreeCollectorBase() =
```
[the remark this answers is not in the document]
Comment (dawedawe) [reply src/FSharp.Analyzers.SDK/TASTCollecting.fs:9]: Yes, I ported all GR analyzers and all tests still pass.

[the remark this answers is not in the document]
Comment (nojaf) [reply src/FSharp.Analyzers.SDK/TASTCollecting.fs:9]: Great!

File: docs/content/Dual Analyzer.fsx [hunk 5]
```diff
@@ -115,6 +115,48 @@ let editorAnalyzer (ctx: EditorContext) : Async<Message list> =
...
 ⚠️ Please do not be tempted by calling `.Value` on the `EditorContext` 😉.
 
+To enable a wide range of analyzers, both context types give access to very detailed information about the source code.  
+Among this information is the full untyped abstract syntax tree (AST) and the typed abstract syntax tree (TAST).
+As you can deduce from the example above, processing these trees is a very common task in an analyzer. But writing your own tree traversal code can be daunting and can also get quite repetitive over many analyzers.  
+That's why the SDK offers the `ASTCollecting` and `TASTCollecting` modules. In there, you'll find facility types and functions to make your analyzers author life easier.
+For both trees, a type is defined, `SyntaxCollectorBase` and `TypedTreeCollectorBase` respectively, with members you can override to have easy access to the tree elements you want to process.  
+Just pass an instance with your overriden members to the `walkAst` or `walkTast` function.  
+
+The `allOpenStatements` function from above, rewritten to make use of `walkAst`, could look like this:
```
Comment (nojaf) [inline docs/content/Dual Analyzer.fsx:125]: I would not show the old code in the first place.
Just straightly go with the new code and explain where these Collecting types are coming from.

File: docs/content/Dual Analyzer.fsx [hunk 6]
```diff
@@ -115,6 +115,48 @@ let editorAnalyzer (ctx: EditorContext) : Async<Message list> =
 Both analyzers will follow the same code path: the console application will always have the required data, while the editor needs to be more careful.  
 ⚠️ Please do not be tempted by calling `.Value` on the `EditorContext` 😉.
 
+To enable a wide range of analyzers, both context types give access to very detailed information about the source code.  
+Among this information is the full untyped abstract syntax tree (AST) and the typed abstract syntax tree (TAST).
+As you can deduce from the example above, processing these trees is a very common task in an analyzer. But writing your own tree traversal code can be daunting and can also get quite repetitive over many analyzers.  
+That's why the SDK offers the `ASTCollecting` and `TASTCollecting` modules. In there, you'll find facility types and functions to make your analyzers author life easier.
+For both trees, a type is defined, `SyntaxCollectorBase` and `TypedTreeCollectorBase` respectively, with members you can override to have easy access to the tree elements you want to process.
```
Comment (nojaf) [inline docs/content/Dual Analyzer.fsx:122]: Add links to API docs: http://localhost:8901/reference/fsharp-analyzers-sdk-astcollecting-syntaxcollectorbase.html

**Review by nojaf — CHANGES_REQUESTED** [review]
Good stuff!

**Review by nojaf — APPROVED** [review]

**Pull request merged**